File: hello-hunter-dynamic-dialer-weekly-reload/functions/zoho-crm-api.js
const axios = require('axios');

class ZohoCrmApi {
    constructor() {
        this.conn = axios.create({
            baseURL: 'https://www.zohoapis.com/crm/v5/',
            headers: { 'Content-Type': 'application/json' },
        });
    }

    async refreshAccessTokens() {
        const refreshUrl = `https://accounts.zoho.com/oauth/v2/token?refresh_token=${process.env.REFRESH_TOKEN}&client_id=${process.env.CLIENT_ID}&client_secret=${process.env.CLIENT_SECRET}&grant_type=refresh_token`;

        try {
            const refreshRequest = await axios.post(refreshUrl);
            const token = refreshRequest.data;

            if (!token || refreshRequest.status === 401 || token.status === 'failure') {
                console.error('Zoho CRM refresh Token not received');
                console.debug(refreshRequest.data);
                console.debug(refreshRequest.status);
                return;
            }

            console.info('Zoho CRM refresh Token received');
            this.conn.defaults.headers['Authorization'] = `Zoho-oauthtoken ${token.access_token}`;
        } catch (error) {
            console.error('Error refreshing Zoho CRM tokens:', error.message);
            if (error.response) {
                console.debug(error.response.data);
                console.debug(error.response.status);
            }
        }
    }

    async search(module, params) {
        try {
            const response = await this.apiCall(`${module}/search?criteria=${params}`, 'get');
            return response.data;
        } catch (error) {
            console.error('Error searching Zoho CRM:', error.message);
            return null;
        }
    }

    async coql(selectQuery) {
        return this.apiCall('coql', 'post', { select_query: selectQuery });
    }

    async zohoCrmPost(url, type, payload) {
        const payloadLen = payload.length;
        const numOfRequests = Math.ceil(payloadLen / 100);
        let startIndex = 0;
        let endIndex = payloadLen > 100 ? 99 : payloadLen - 1;
        const respArray = [];

        for (let iteration = 1; iteration <= numOfRequests; iteration++) {
            const data = { data: payload.slice(startIndex, endIndex + 1) };
            const apiResp = await this.apiCall(url, type, data);

            if (apiResp?.data) {
                respArray.push(apiResp.data);
            }

            startIndex += 100;
            endIndex = Math.min(startIndex + 99, payloadLen - 1);
        }

        return respArray;
    }

    async apiCall(url, type, payload = null) {
        try {
            const response = await this.conn.request({
                url,
                method: type,
                data: payload ? JSON.stringify(payload) : null,
            });

            switch (response.status) {
                case 200:
                case 201:
                case 202:
                    console.info(`Zoho ${type} to ${url} successful`);
                    return response.data;
                case 204:
                    console.info(`Zoho ${type} to ${url} successful`);
                    console.info(response.data);
                    return 'no data';
                default:
                    console.error(`Zoho ${type} to ${url} failed`);
                    console.debug(response.data);
                    console.debug(response.status);
                    return 'fail';
            }
        } catch (error) {
            console.error(`Error making API call to ${url}:`, error.message);
            if (error.response) {
                console.debug(error.response.data);
                console.debug(error.response.status);
            }
            return 'fail';
        }
    }
}

module.exports = ZohoCrmApi;

File: Ruby/create-weekly-scrub-batch-invoices-a2a1e597-62f8-44b7-848b-a452395c6a16/helpers/zoho-crm-api.rb
class ZohoCrmApi
  attr_accessor :conn, :headers

  def initialize
    self.conn = Faraday.new 'https://www.zohoapis.com/crm/v5/'
    self.refresh_access_tokens
    self
  end

  def refresh_access_tokens
    refresh_url = "https://accounts.zoho.com/oauth/v2/token?refresh_token=#{ENV['REFRESH_TOKEN']}&client_id=#{ENV['CLIENT_ID']}&client_secret=#{ENV['CLIENT_SECRET']}&grant_type=refresh_token"
    refresh_request = @conn.post refresh_url
    token = JSON.parse(refresh_request.body)
    if token == nil || refresh_request.status == 401 || token['status'] == 'failure'
      $logger.error('Zoho CRM refresh Token not received')
      $logger.debug(refresh_request.body)
      $logger.debug(refresh_request.status)
    else
      $logger.info('Zoho CRM refresh Token received')
    end
    @conn.headers['Authorization'] = "Zoho-oauthtoken #{token['access_token']}"
  end

  def search(z_module, params)
    response = self.api_call("#{z_module}/search?criteria=#{params}", 'get')
    response['data']
  end

  def coql(select_query)
    self.api_call('coql', 'post', {select_query: select_query})
  end

  def zoho_crm_post(url, type, payload)
    payload_len = payload.length
    num_of_requests = (payload_len.to_f / 100 ).ceil
    start_index = 00
    end_index =
      if payload_len > 100
        99
      else
        payload_len - 1
      end
    resp_array = []

    for iteration in 1..num_of_requests do
      data = {data: payload[start_index..end_index]}
      api_resp = self.api_call(url, type, data)
      resp_array.append(api_resp['data'])
      start_index += 100
      end_index += 100
      next_iter = iteration + 1
      end_index = payload_len - 1 if next_iter == num_of_requests
    end
    resp_array
  end

  def api_call(url, type, payload = nil)
    json_payload = payload.nil? ? payload : JSON.generate(payload)
    $logger.info("#{url} #{json_payload}")
    response =
      case type
      when 'get' then @conn.get(url,json_payload)
      when 'post' then @conn.post(url,json_payload)
      when 'put' then @conn.put(url,json_payload)
      end
    
    $logger.info
    case response.status
    when 200, 201, 202
      data = JSON.parse(response.body)
      $logger.info("Zoho #{type} to #{url} successful")
      return data
    when 204
      $logger.info("Zoho #{type} to #{url} successful")
      $logger.info(response.body)
      return 'no data'
    else
      $logger.error("Zoho #{type} to #{url} failed")
      $logger.debug(response.body)
      $logger.debug(response.status)
      return 'fail'
    end
  end
end
